source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,553 @@
+
+  // The new condition block.
+  ir::BasicBlock* new_condition_block;
+
+  // The new header block.
+  ir::BasicBlock* new_header_block;
+
+  // A mapping of new block ids to the original blocks which they were copied
+  // from.
+  std::map<uint32_t, ir::BasicBlock*> new_blocks;
+
+  // A mapping of new instruction ids to the instruction ids from which they
+  // were copied.
+  std::map<uint32_t, uint32_t> new_inst;
+};
+
+class LoopUnrollerUtilsImpl {
+ public:
+  LoopUnrollerUtilsImpl(ir::IRContext* c, ir::Function& function)
+      : ir_context_(c), function_(function) {}
+
+  void RemapResultIDs(ir::Loop* loop, ir::BasicBlock* BB);
+
+  void RemapOperands(ir::BasicBlock* BB,
+                     std::map<uint32_t, uint32_t>& new_inst);
+
+  void CopyBody(ir::Loop* loop, bool eliminate_conditions);
+
+  void PartiallyUnroll(ir::Loop*, int);
+
+  void PartiallyUnrollUnevenFactor(ir::Loop* loop, int factor);
+
+  void FullyUnroll(ir::Loop* loop);
+
+  uint32_t GetPhiVariableID(const ir::Instruction* phi, uint32_t label) const;
+  void CloseUnrolledLoop(ir::Loop* loop);
+  void FoldConditionBlock(ir::BasicBlock* condtion_block, uint32_t new_target);
+
+  void AddBlocksToFunction(const ir::BasicBlock* insert_point);
+
+  ir::Loop DuplicateLoop(ir::Loop* old_loop);
+
+  void CopyBasicBlock(ir::Loop* loop, const ir::BasicBlock* itr);
+
+ private:
+  ir::IRContext* ir_context_;
+
+  ir::Function& function_;
+  LoopUtils::BasicBlockListTy blocks_to_add_;
+  LoopStateUtil state_;
+  void Unroll(ir::Loop*, int);
+};
+
+void LoopUnrollerUtilsImpl::PartiallyUnrollUnevenFactor(ir::Loop* loop,
+                                                        int factor) {
+  // Create a new merge block for the first loop.
+  std::unique_ptr<ir::Instruction> new_label{new ir::Instruction(
+      ir_context_, SpvOp::SpvOpLabel, 0, ir_context_->TakeNextId(), {})};
+  std::unique_ptr<ir::BasicBlock> new_exit_bb{
+      new ir::BasicBlock(std::move(new_label))};
+
+  // Save the id of the block before we move it.
+  uint32_t new_merge_id = new_exit_bb->id();
+
+  // Add the block the the list of blocks to add, we want this merge block to be
+  // right at the start of the new blocks.
+  blocks_to_add_.push_back(std::move(new_exit_bb));
+
+  // Duplicate the loop, providing access to the blocks of both loops.
+  ir::Loop new_loop = DuplicateLoop(loop);
+
+  // Make the first loop branch to the second.
+  std::unique_ptr<ir::Instruction> new_branch{new ir::Instruction(
+      ir_context_, SpvOp::SpvOpBranch, 0, 0,
+      {{SPV_OPERAND_TYPE_ID, {new_loop.GetHeaderBlock()->id()}}})};
+  blocks_to_add_[0]->AddInstruction(std::move(new_branch));
+
+  Unroll(&new_loop, factor - 1);
+
+  analysis::Integer uint(32, false);
+  uint32_t uint32_type_id =
+      ir_context_->get_type_mgr()->GetTypeInstruction(&uint);
+
+  // We need to account for the initial body when calculating the remainder.
+  uint32_t remainder = loop->NumIterations() % (factor + 1);
```

Comment:
that should be `factor`, if `loop->NumIterations() % (factor + 1)` returns factor it means you could have unrolled with this pre-loop.